File: Algorithms/Python/45 Jump Game II.py
# 1. out of time
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        min_jumps = list(range(n))
        for i in range(n):
            for j in range(i):
                if (j + nums[j] >= i):
                    min_jumps[i] = min(min_jumps[j] + 1, min_jumps[i])
        
        return min_jumps[-1]
        
        
        
# 2. greedy
class Solution:
    def jump(self, nums: List[int]) -> int:
        position = len(nums) - 1
        step = 0

        while position > 0:
            i = 0
            while i < position:
                if nums[i] + i >= position:
                    position = i
                    step += 1
                
                i += 1

        return step
        

File: Algorithms/C/45 Jump Game II.c
// 1 greedy
int jump(int* nums, int numsSize){
    int position = numsSize - 1;
    int step = 0;
    while (position > 0) {
        for (int i = 0; i < position; ++i) {
            if (nums[i] + i >= position) {
                position = i;
                ++step;
            }
        }
    }

    return step;
}

// 2. O(n^2)
int jump(int* nums, int numsSize){
    int rightmost = 0;
    int end = 0;
    int step = 0;

    for (int i = 0; i < numsSize - 1; ++i) {
        rightmost = fmax(rightmost, nums[i] + i);
        if (i == end) {
            end = rightmost;
            ++step;
        }
    }

    return step;
}
